File: completed-labs/06/cpp/Stack.cpp
#include "SLList.hpp"
#include "Stack.hpp"

using namespace std;

void Stack::push(int o) {
  head = new SLList(o, head);
}

int Stack::pop() {
  int r = head->first();
  SLList *newhead = head->rest();
  delete head;
  head = newhead;
  return r;
}

int Stack::top() {
  return head->first();
}

bool Stack::empty() {
  return head == SLList::NIL;
}

File: is52038b-labs/06/java/Stack.java
public class Stack {
    private SLList head = SLList.NIL;
    public Stack() { }
    public void push(Object o) {
	head = new SLList(o, head);
    }
    public Object pop() {
	Object r = head.first();
	head = head.rest();
	return r;
    }
    public Object top() {
	return head.first();
    }
    public boolean empty() {
	return head == SLList.NIL;
    }
}
